dsc/workflows/opencourseware.py

```diff
@@ -0,0 +1,203 @@
+    """Workflow for OpenCourseWare (OCW) deposits.
+
+    The deposits managed by this workflow are requested by
+    Technical Services staff and are for submission to DSpace@MIT.
+    """
+
+    workflow_name: str = "opencourseware"
+    metadata_mapping_path: str = "dsc/workflows/metadata_mapping/opencourseware.json"
+
+    def reconcile_bitstreams_and_metadata(self) -> tuple[set[str], set[str]]:
+        """Reconcile bitstreams against metadata.
+
+        Generate a list of bitstreams without item identifiers and item identifiers
+        without bitstreams. Any discrepancies will be addressed by the engineer and
+        stakeholders as necessary.
+
+        For OpenCourseWare deposits, the zip files are the bitstreams to be deposited
+        into DSpace, but they also must contain a 'data.json' file, which is the metadata.
+        As such, the 'reconcile' method only determines whether there are any
+        bitstreams without metadata (any zip files without a 'data.json').
+        Metadata without bitstreams are basically impossible because the metadata
+        ('data.json') is inside the bitstream (zip file), hence an empty set is returned.
+        """
+        bitstream_dict = self._build_bitstream_dict()
+
+        # extract item identifiers from bitstream_dict
+        item_identifiers = list(bitstream_dict.keys())
+
+        bitstreams_without_metadata = set(item_identifiers) - set(
+            self._identify_bitstreams_with_metadata(item_identifiers)
+        )
+        item_metadata_without_bitstreams: set[str] = set()
+        return item_metadata_without_bitstreams, bitstreams_without_metadata
+
+    def _build_bitstream_dict(self) -> dict:
+        """Build dictionary of bitstreams.
+
+        This method will look for zip files within the designated 'batch' folder
+        of the S3 bucket (i.e., self.batch_path). In the case of OpenCourseWare
+        deposits, this method expects:
+            * a single (1) zipped file per item identifier
+            * filename of zipped file must correspond to item identifier
+              (i.e., '<item_identifier>'.zip)
+        """
+        s3_client = S3Client()
+        bitstreams = list(
+            s3_client.files_iter(
+                bucket=self.s3_bucket, prefix=self.batch_path, file_type=".zip"
+            )
+        )
+        bitstream_dict: dict[str, list[str]] = defaultdict(list)
+        for bitstream in bitstreams:
+            file_name = bitstream.split("/")[-1]
+            item_identifier = file_name.replace(".zip", "")
+            bitstream_dict[item_identifier].append(bitstream)
+        return bitstream_dict
+
+    def _identify_bitstreams_with_metadata(
+        self, item_identifiers: list[str]
+    ) -> list[str]:
+        bitstreams_with_metadata = []
+        for item_identifier in item_identifiers:
+            try:
+                zip_file = (
+                    f"s3://{self.s3_bucket}/{self.batch_path}/{item_identifier}.zip"
+                )
```

Comment:
What if `self._build_bitstream_dict()` returned a tuple like `(<indentifier>, <bitstream_filename>, <bitstream_fullpath>)`?  If you had the full path of the bitstream returned there, when discovered, you could theoretically pass that list of tuples to _this_ method and it wouldn't have to construct the path here.

I certainly see how this works, but feels potentially brittle and/or confusing why we use `s3_client.files_iter(...)` with parameters in one place, but we manually construct S3 URIs here.

You could do something like:
```python
for item_identifier, _, bitstream_uri in bitstreams:
   # do stuff
```

where `bitstreams` is this slightly modified list of bitstreams in S3 that were discovered via `self._build_bitstream_dict()`.

Reply:
Please review the [latest version of `OpenCourseWare.reconcile_bitstreams_and_metadata`](https://github.com/MITLibraries/dspace-submission-composer/blob/IN-1096-ocw-workflow/dsc/workflows/opencourseware.py#L27-L62) and let me know what you think! Changes introduced in PR #102 allowed more flexibility when defining the 'reconcile' method for OCW.